### ISSUE TITLE
Question: pack an array-of-struct as a struct-of-array?

### ISSUE THREAD

**Issue by wenyan4work** [msgpack/msgpack-c#986]
msgpack-c helped us a lot for computing data storage. One of the most frequent use case is like this:
```cpp
class myclass
{
private:
    long gid{5};
    std::array<double, 3> m_vec{1, 2, 3};

public:
    MSGPACK_DEFINE_MAP(gid, m_vec);
    // MSGPACK_DEFINE(m_str, m_vec);
};

std::vector<myclass> vec(5);
msgpack::sbuffer sbuf;
msgpack::pack(sbuf, vec);
```
Then the packed data is like this:
```python
[{'gid': 5, 'm_vec': [1.0, 2.0, 3.0]}, {'gid': 5, 'm_vec': [1.0, 2.0, 3.0]}, {'gid': 5, 'm_vec': [1.0, 2.0, 3.0]}, {'gid': 5, 'm_vec': [1.0, 2.0, 3.0]}, {'gid': 5, 'm_vec': [1.0, 2.0, 3.0]}]
```
The names of member variables are saved for every struct in this `std::vector<myclass>`. This is a lot of overhead. Using `MSGPACK_DEFINE` helps, but the bytes for type information is still redundant.

Then, what is the easiest way to save this `std::vector<myclass>` in the struct-of-array style? One way is to write a specialization of pack<> to pack `std::vector<myclass>` manually. Another way is to write a friend class like this
```
class myclass_vec
{
private:
    std::vector<long> gid;
    std::vector<std::array<double, 3>> m_vec;

public:
    myclass_vec(const std::vector<myclass> &vec)
    {
        for (const auto &x : vec)
        {
            gid.push_back(x.gid);
            m_vec.push_back(x.m_vec);
        }
    }
    MSGPACK_DEFINE_MAP(gid, m_vec);
};
```
and saved data is like this
```python
{'gid': [5, 5, 5, 5, 5], 'm_vec': [[1.0, 2.0, 3.0], [1.0, 2.0, 3.0], [1.0, 2.0, 3.0], [1.0, 2.0, 3.0], [1.0, 2.0, 3.0]]}
```
This friend class does the trick, but if  `myclass` contains another customized class this might get complicated, too. 

In short, what is the best practice to pack an array-of-struct as a struct-of-array?
